addons/web_editor/tools.py

```diff
@@ -0,0 +1,133 @@
+            return ('youku', youku_link, youku_match)
+    return False
+
+
+def get_video_url_data(video_url, **kwargs):
+    ''' Computes the platform name and embed_url from given URL
+        (or error message in case of invalid URL).
+    '''
+    source = get_video_source_data(video_url)
+    if not source:
+        return {'error': True, 'message': _('The provided url is invalid')}
+
+    autoplay = kwargs.get('autoplay') and '?autoplay=1&mute=1' or '?autoplay=0'
+    controls = kwargs.get('hide_controls') and '&controls=0' or ''
+    loop = kwargs.get('loop') and '&loop=1' or ''
+
+    # We directly use the provided URL as it is
+    embed_url = video_url
+    platform = source[0]
+    platform_id = source[1]
+    platform_match = source[2]
+    if platform == 'youtube' and len(platform_id) == 11:
+        fullscreen = kwargs.get('hide_fullscreen') and '&fs=0' or ''
+        youtube_loop = loop and loop + '&playlist=%s' % (platform_id) or ''
+        logo = kwargs.get('hide_yt_logo') and '&modestbranding=1' or ''
+        params = '%s&rel=0%s%s%s%s' % (autoplay, youtube_loop, controls, fullscreen, logo)
+        embed_url = '//www.youtube%s.com/embed/%s%s' % (platform_match.groups()[0] or '', platform_id, params)
+    elif platform == 'vimeo':
+        vimeo_autoplay = autoplay.replace('mute', 'muted')
+        embed_url = '//player.vimeo.com/video/%s%s%s' % (platform_id, vimeo_autoplay, loop)
+    elif platform == 'dailymotion':
+        logo = kwargs.get('hide_dm_logo') and '&ui-logo=0' or ''
+        share = kwargs.get('hide_dm_share') and '&sharing-enable=0' or ''
+        params = '%s%s%s%s' % (autoplay, controls, logo, share)
+        embed_url = '//www.dailymotion.com/embed/video/%s%s' % (platform_id, params)
+    elif platform == 'instagram':
+        embed_url = '//www.instagram.com/p/%s/embed/' % (platform_id)
+    elif platform == 'youku':
+        embed_url = '//player.youku.com/embed/%s' % (platform_id)
+
+    return {'platform': platform, 'embed_url': embed_url}
+
+
+def get_video_embed_code(video_url):
+    ''' Computes the valid iframe from given URL that can be embedded
+        (or False in case of invalid URL).
+    '''
+    data = get_video_url_data(video_url)
+    if data.get('error'):
+        return False
+    return '<iframe class="embed-responsive-item" src="%s" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowFullScreen="true" frameborder="0"></iframe>' % data.get('embed_url')
+
+
+def get_video_thumbnail(video_url):
+    ''' Computes the valid thumbnail image from given URL
+        (or False in case of invalid URL).
+    '''
+    source = get_video_source_data(video_url)
+    if not source:
+        return False
+
+    response = False
+    platform = source[0]
+    platform_id = source[1]
+    if platform == 'youtube' and len(platform_id) == 11:
+        response = requests.get('https://img.youtube.com/vi/' + platform_id + '/0.jpg')
```

Comment:
This should go inside a standalone test if you were to keep calling third party services during test.
This will create race condition and makes build fail sometimes.
You can grep @standalone('cow_views_inherit') to see how it works.

Pro: It will not create race condition
Con: It will only be tested during the nightly, except if @d-fence can confirm standalone tests are now also run on every build (at least every staging) except the standalone test from ADT about module operations? If not, that should be.

But you don't have a choice to put it in standalone anyway :D

Reply:
Alright, I'll do it

Reply:
Whoops, my bad.
Standalone tests are only for tests doing some module operations (install etc).
Tests doing external calls should be regular tests with `@tagged('-standard', 'external')`. Those will only be run in the nightly build.